django/db/backends/postgresql/base.py

```diff
@@ -21,47 +21,98 @@
+    TSTZRANGE_OID = psycopg.postgres.types["tstzrange"].oid
+
+    # RemovedInDjango51Warning.
+    class CIStr(str):
+        pass
+
+else:
+    import psycopg2.extensions
+    import psycopg2.extras
+
+    psycopg2.extensions.register_adapter(SafeString, psycopg2.extensions.QuotedString)
+    psycopg2.extras.register_uuid()
+
+    # Register support for inet[] manually so we don't have to handle the Inet()
+    # object on load all the time.
+    INETARRAY_OID = 1041
+    INETARRAY = psycopg2.extensions.new_array_type(
+        (INETARRAY_OID,),
+        "INETARRAY",
+        psycopg2.extensions.UNICODE,
+    )
+    psycopg2.extensions.register_type(INETARRAY)
+    CIStr = str
+
+# Some of these import psycopg, so import them after checking if it's installed.
+from .client import DatabaseClient  # NOQA isort:skip
+from .creation import DatabaseCreation  # NOQA isort:skip
+from .features import DatabaseFeatures  # NOQA isort:skip
+from .introspection import DatabaseIntrospection  # NOQA isort:skip
+from .operations import DatabaseOperations  # NOQA isort:skip
+from .psycopg_any import IsolationLevel, is_psycopg3  # NOQA isort:skip
+from .schema import DatabaseSchemaEditor  # NOQA isort:skip
+
 
-# Some of these import psycopg2, so import them after checking if it's installed.
-from .client import DatabaseClient  # NOQA
-from .creation import DatabaseCreation  # NOQA
-from .features import DatabaseFeatures  # NOQA
-from .introspection import DatabaseIntrospection  # NOQA
-from .operations import DatabaseOperations  # NOQA
-from .schema import DatabaseSchemaEditor  # NOQA
+@lru_cache
+def get_adapters_template(use_tz, timezone):
+    # Create at adapters map extending the base one to base connections on
+    ctx = psycopg.adapt.AdaptersMap(psycopg.adapters)
 
-psycopg2.extensions.register_adapter(SafeString, psycopg2.extensions.QuotedString)
-psycopg2.extras.register_uuid()
+    # Register a no-op dumper to avoid a round trip from psycopg version 3
+    # decode to json.dumps() to json.loads(), when using a custom decoder
+    # in JSONField.
+    ctx.register_loader("jsonb", TextLoader)
 
-# Register support for inet[] manually so we don't have to handle the Inet()
-# object on load all the time.
-INETARRAY_OID = 1041
-INETARRAY = psycopg2.extensions.new_array_type(
-    (INETARRAY_OID,),
-    "INETARRAY",
-    psycopg2.extensions.UNICODE,
-)
-psycopg2.extensions.register_type(INETARRAY)
+    # Don't convert automatically from Postgres network types to Python ipaddress
+    ctx.register_loader("inet", TextLoader)
+    ctx.register_loader("cidr", TextLoader)
+    ctx.register_dumper(Range, DjangoRangeDumper)
+
+    # Register a timestamptz loader configured on self.timezone.
+    # This, however, can be overridden by create_cursor.
+    register_tzloader(timezone, ctx)
+
+    # Dump strings using the text oid, where the default unknown oid
+    # doesn't work well (e.g. in variadic functions)
+    ctx.register_dumper(str, StrDumper)
```

Comment:
I have commented [here](https://github.com/django/django/commit/3caa2c51b8f762cc3c260dddb6a60a0e1c9621b8#r90534349) why this is a bad idea. I don't know if that message has only reached the author of the commit: it was the wrong place to start that conversation.

Reply:
https://github.com/apollo13/django/pull/4 should address your concerns @dvarrazzo. I have reservations about the arguments provided in the thread in a context where Django has always configured its psycopg connections in a way to make them work with its public ORM API first and that usage of custom types and raw queries will have to adapted anyway by the advanced users to migrate to psycopg 3.

Reply:
I am somewhat torn. I think registering `StrDumper` for `str` makes it indeed easier for the ORM but I also tend to agree with Daniele that `unknown` is probably the better default. We learned a lot though and I will just remove the `StrDumper` again in my branch and see which tests would need fixing.

Reply:
Ok, I will mark this one as resolved since we dropped `StrDumper` again.

Reply:
Was it a big work to fix the tests after reverting? I'm genuinely curious because this choice of types adaptation was something [extremely delicate](https://www.varrazzo.com/blog/2020/11/07/psycopg3-adaptation/) to do in psycopg 3 in order to make the system behave the way people expects...

Reply:
Not really, it was mainly variadic functions:
https://github.com/django/django/pull/15687/commits/f29c66102d38b19d166d13afec040ea60809489e

and fixing `%s IS NULL` for unknown:
https://github.com/django/django/pull/15687/commits/2ad89f05995bfaf999631f65b6b9b02bd665e115

There might be a thing here or there where we already fixed it in the past, but those to where basically the last ones (in addition to send json as jsonb to the database).

Reply:
The `IS NULL` case is really ugly, and seems really unnecessary: null is a null whatever the type :(

```python
>>> conn.execute("select %s is not distinct from null", ['asdf']).fetchone()
IndeterminateDatatype: could not determine data type of parameter $1
```

Maybe it might be fixed by postgres developers, but even in that case the kludge will have to remain in order to support existing databases. It might be even worth hearing from them to figure out if there is an alternative syntax working for unknown too.